### ISSUE TITLE
[QUESTION] How to deal with logical types other than decimal?

### ISSUE THREAD

**Issue by rocketraman** [commercehub-oss/gradle-avro-plugin#54]
The Avro schema defines several logical types, for example, timestamp-millis. As I understand it, these types are supposed to be converted to Joda-Date types when the schema is compiled. However, when compiling a schema like this:

```
{
  "namespace": "myns",
  "type": "record",
  "name": "Foo",
  "fields": [
    {"name": "timestamp", "type": "long", "logicalType": "timestamp-millis"}
  ]
}
```

the generated output contains only a `long` field.

What I really want to do here is register a `Conversion` that converts `timestamp-millis` into a java.time `Instant`, since joda-date is deprecated, and ideally have my generated code get and set `Instant`s. Is there a way to achieve this? The Avro docs on this are pretty sparse...

**Comment by rocketraman** [comment]
Ok, figured out the first part of my question: looks like my schema was wrong. The timestamp field should have been defined as:

`{"name": "timestamp", "type": {"type": "long", "logicalType": "timestamp-millis"}}`

which then does cause the compiler to emit a Joda `DateTime` field.

So then, to the second part of my question: can I hook into the compiler process somehow to change the output of a logicalType from a Joda `DateTime` to a java.time `Instant`?

**Comment by rocketraman** [comment]
@davidmc24 I've seen that post, and already commented on the relevant issue :-) But it seems like the Avro maintainers are extremely inactive / non-responsive, so I was wondering if there was something that could be done at the plugin level to workaround this gap. I have the converters defined as per one of the StackOverflow answers, but I guess would need to register them with the compiler somehow, via the avro plugin.

**Comment by davidmc24** [comment]
So... we're in the realm of hacky workarounds at this point.  If I was in your situation and needed it before the Avro maintainers publish a release, I'd consider:

* Setting up a https://jitpack.io/ repository to get access to avro binaries based on the pull request
* Temporarily fork this plugin and make the following changes:
  * Depend on the jitpack binaries rather than the official avro binaries for compilation
  * Add support for the new `dateTimeLogicalTypeType` setting in SpecificCompiler
* Update my project to use the forked version of the plugin and the custom avro binary

Alternatively, one could consider hacks outside the plugin.  For example, a gradle task after the tasks from this plugin runs that does a text replacement on the generated files to substitute the desired types.

**Comment by rocketraman** [comment]
Just to close the loop on this, with Avro 1.9.0 getting close to release, I've created pull https://github.com/commercehub-oss/gradle-avro-plugin/pull/70 to support 1.9.0 and JSR310 types.

**Comment by tyocum17** [comment]
I know this issue was closed and it addresses the Java 8 dates. I'm trying to figure out how to register a custom conversion for OffsetDateTime that currently isn't supported. I created a conversion class:

`public class OffsetDateTimeConversion extends Conversion<OffsetDateTime>`

with logicalTypeName = 'offset-date-time' that implements `toCharSequence` and 'fromCharSequence` so I can convert to/from ISO 8601 date/time string with timezone offset.

What I can't figure out is how to register the "offset-date-time" logical type. Debugging the maven-plugin, I see when it parses my schema that contains a field with a custom logical type:
`
"name": "createDateTime",
            "type": {
                "type": "string",
                "logicalType": "offset-date-time"
            }
`
it tries to set the field schema's logicalType property and since it isn't a built-in one ("timestamp-millis", "decimal", "date", etc) it looks for it in the REGISTERED_TYPES map but it doesn't exist.

The result is my generated class doesn't contain the conversion for my custom logical type,

The code I need to execute to register the logical type is:
`
LogicalTypes.register(getLogicalTypeName(), schema -> new OffsetDateTimeLogicalType());
`

Without modifying plugin code, I don't see a why to register my logical type before the schema parses. Just wondering if you have any experience/luck with this. Thanks!
